### PR TITLE
DO-NOT-MERGE Add implementation of ipython widget

### PR DESCRIPTION
**Pull request by chmp**
As discussed in https://github.com/vega/ipyvega/issues/93, this PR contains an implementation for an IPython Vega widget. In addition to the existing display system, it allows

- to put altair charts in a larger UI with other widgets,
- to change the chart without re-displaying the widget, and
- to dynamically update the data via changesets.

I struggled a bit with webpack to get it to only optionally define the widget component. Not sure whether my solution is idiomatic.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -92,3 +92,56 @@ export function render(
     })
     .catch(error => showError(el, error));
 }
+
+// optionally define a Jupyter Widget if the plugin is available available
```
Comment (domoritz) [inline src/index.ts]: double available

File: src/index.ts [hunk 2]
```diff
@@ -92,3 +92,56 @@ export function render(
...
+export const VegaWidget = import('@jupyter-widgets/base')
+.then(widgets => {
+  return class VegaWidget extends widgets.DOMWidgetView {
+    view: any
+
+    render() {
+          const reembed = () => {
```
Comment (domoritz) [inline src/index.ts]: please use 2 spaces

File: vega/widget.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    """An IPython widget display a vega chart.
+
+    Specifying the spec directly::
+
+        widget = VegaWidget({...})
+        widget.update(remove='datum.t < 5', insert=[{...}, {...}])
+
+    Usage with ``altair``::
+
+        widget = VegaWidget.from_cart(chart)
```
Comment (domoritz) [inline vega/widget.py]: from cart, lol ;-)

File: src/index.ts [hunk 4]
```diff
@@ -92,3 +92,56 @@ export function render(
...
+              }
+              if(this.view == null) {
+                  console.error('no view attached to widget');
+                  return
+              }
+
+              const filter = new Function('datum', 'return (' + (ev.remove || 'false') + ')');
+              const newValues = ev.insert || [];
+
+              const changeSet = (
```
Comment (domoritz) [inline src/index.ts]: why not just 

```ts
const changeSet = this.view.changeset().insert(newValues).remove(filter)
```

You can access `vegaEmbed.vega.changeset()` as well but this is fine.

File: src/index.ts [hunk 5]
```diff
@@ -92,3 +92,56 @@ export function render(
     })
     .catch(error => showError(el, error));
 }
+
+// optionally define a Jupyter Widget if the plugin is available available
+export const VegaWidget = import('@jupyter-widgets/base')
+.then(widgets => {
+  return class VegaWidget extends widgets.DOMWidgetView {
+    view: any
```
Comment (domoritz) [inline src/index.ts]: make this `View` from Vega

**Comment by chmp** [comment]
@domoritz my yarn upgraded me to `vega-embed==3.16.0` and now `Mode` is no longer available for import. Could you give a pointer how to update the import in the first line of `index.ts`? Thanks. 

Once I get it to compile again, I will push commits addressing your points above :).

**Comment by domoritz** [comment]
Ahh, let me fix that in Embed.

**Comment by domoritz** [comment]
Sorry about that. Please update to 3.16.1.

**Comment by domoritz** [comment]
Can you rebase against master? this will update all the dependencies.

**Comment by chmp** [comment]
Hm. I have to backtrack here. I used a new venv to check everything and it seems that the optional dependency part within webpack is not working as expected in production mode. Now, it seems extensions are loaded in a different order and the whole vega extension is no longer loaded.

**Pull request closed without merging**